Switching to DataContractJsonSerializer

danielmarbach opened the pull request:
Since we plan to internalize the serializers the custom checks now use the built in DataContractJsonSerializer

@Particular/servicecontrol-maintainers please review

cc @andreasohlund @mat-mcloughlin

andreasohlund commented:
Should we add a acceptance test verifying this? (just running a plain endpoint with the json serializer to simulate SC to check that we can handle it)

danielmarbach commented:
@andreasohlund I did it manually and diffed the plain json output like we discussed today on the call. If you think we should add that test I can do it. Don't know how much value this test provides because I would essentially need to simulate assumptions of SC into that test.

andreasohlund commented:
@danielmarbach I meant more in general, it seems that the only end to end test we have here is a sample relying on someone hitting F5 now and then. I'd say we should add a few acpt tests? (can be a separate pull)

danielmarbach commented:
@andreasohlund @SzymonPobiega @WilliamBZA Added the necessary acceptance tests and deleted the samples

danielmarbach commented:
Acceptance Tests on build server setup

andreasohlund commented on src/ServiceControl.Plugin.Nsb6.CustomChecks.AcceptanceTests/ScenarioDescriptors/Persistence.cs:
```diff
@@ -0,0 +1,92 @@
+﻿namespace NServiceBus.AcceptanceTests.ScenarioDescriptors
```
Do we need all this given that we won't change persistence? (same for serializers etc etc)

Can we go with a "hardcoded" default server instead?

andreasohlund commented:
The tests seems to run on the build server but why is the branch reported as `master` ?

http://builds.particular.net/viewLog.html?buildId=143126&tab=buildResultsDiv&buildTypeId=ServiceControl_ServiceControl_Plugin_Nsb6_CustomChecks_2AcceptanceTests

johnsimons commented:
I'm sure this was discussed but why not use Json.net since that is the lib we use in SC to deserialize the payload ?
Wouldn't that be the safer way going forward.

danielmarbach commented:
@johnsimons

As far as I understand the code you are using a message handler

https://github.com/Particular/ServiceControl/blob/master/src/ServiceControl/CustomChecks/ReportCustomCheckResultHandler.cs#L16

and the JsonSerializer on the endpoint. Since the JsonSerializer is in the
core ILMerged with JSon.NET we would need to reference Json.NET in the
plugins and ILMerge it as well. The DataContractSerializer comes with the
framework. The payload of CustomChecks and Heartbeat is well defined.

johnsimons commented:
@danielmarbach in SC we use the json lib directly, we do not use the JsonSerializer that is shipped in the nsb core.

johnsimons commented:
You are right @danielmarbach, those payloads are so simple that it should be safe to use DataContractJsonSerializer, or even strings :wink: 

Just make sure you test weird characters, that was something that gave us problems and json.lib handled it correctly regarding the encoding.

SzymonPobiega commented:
I think that getting rid of type annotations is key. When we are down to plain json, the choice of serializer is much less important.

SzymonPobiega commented:
Here's what I did for Heartbeats: https://github.com/Particular/ServiceControl.Plugin.Nsb6.Heartbeat/commit/93a892cec717eef7a3a017670a7e6a9fbc7f18a3

I've removed the object array (which removed the need for type annotations crap) and added an acceptance test proving an NSB endpoint with json serialization (an approximation of ServiceControl) can handle such messages.

danielmarbach commented:
The last time I ckecked with SC 1.8.3 it didn't work without the
annotations. That's how I detected the bug in NSB which was fixed in 5.2.14

On Tuesday, February 2, 2016, Szymon Pobiega notifications@github.com
wrote:

> Here's what I did for Heartbeats:
> Particular/ServiceControl.Plugin.Nsb6.Heartbeat@93a892c
> https://github.com/Particular/ServiceControl.Plugin.Nsb6.Heartbeat/commit/93a892cec717eef7a3a017670a7e6a9fbc7f18a3
> 
> I've removed the object array (which removed the need for type annotations
> crap) and added an acceptance test proving an NSB endpoint with json
> serialization (an approximation of ServiceControl) can handle such
> messages.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/Particular/ServiceControl.Plugin.Nsb6.CustomChecks/pull/6#issuecomment-178391166
> .

danielmarbach commented:
Ok got it now. I'll try the same for CustomChecks

On Tuesday, February 2, 2016, Daniel Marbach daniel.marbach@particular.net
wrote:

> The last time I ckecked with SC 1.8.3 it didn't work without the
> annotations. That's how I detected the bug in NSB which was fixed in 5.2.14
> 
> On Tuesday, February 2, 2016, Szymon Pobiega <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
> 
> > Here's what I did for Heartbeats:
> > Particular/ServiceControl.Plugin.Nsb6.Heartbeat@93a892c
> > https://github.com/Particular/ServiceControl.Plugin.Nsb6.Heartbeat/commit/93a892cec717eef7a3a017670a7e6a9fbc7f18a3
> > 
> > I've removed the object array (which removed the need for type
> > annotations crap) and added an acceptance test proving an NSB endpoint with
> > json serialization (an approximation of ServiceControl) can handle such
> > messages.
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/Particular/ServiceControl.Plugin.Nsb6.CustomChecks/pull/6#issuecomment-178391166
> > .

danielmarbach commented:
After talking to @andreasohlund @SimonCropp I'm switching back to DataContractSerializer. The data format is well known with custom checks, performance in custom checks is not really the key and having smaller assembly sizes is the bigger win here
